R/statistics.R

```diff
@@ -691,3 +691,247 @@ SecretarySampler <- R6::R6Class("SecretarySampler", public = list(
+        update = function(x, k) {
+            if (length(x) > 1) {
+                stop("Only single-value updated allowed")
+            }
+            private$arm_info[[k]]$mean$update(x)
+            private$n_observed <- private$n_observed + 1
+        }
+    ),
+    active = list(
+        #' @description Returns the the list containing values known to
+        #' to the Bandit in the current state.
+        #'
+        #'
+        #' @return list with summary of the state of the Bandit.
+        value = function() {
+            private$get_state()
+        }
+    ),
+    private = list(
+        K = NULL,
+        T = NULL,
+        N = NULL,
+        which = NULL,
+        n_observed = 0,
+        state = NULL,
+        arm_info = NULL,
+        #' @description Creates and returns a list with information about
+        #' the values known to the Bandit in the current state.
+        #'
+        #' @return list with summary of the state of the Bandit.
+        get_state = function() {
+            arm_means <- c()
+            for (k in seq(K)) {
+                arm_means[k] <- private$arm_info[[k]]$mean$value
+            }
+            state <- list(
+                which = private$which,
+                state = private$state,
+                n_observed = private$n_observed,
+                arm_means = arm_means
+            )
+            state
+        }
+    )
+)
+#' Create a streamer for the uniform exploration strategy in the multi-armed
+#' bandit problem.
+#'
+#' @description \code{ExploreFristNBandit}
+#'
+#' @docType class
+#'
+#' @examples
+#' means <- c(1, 2)
+#' N <- 3
+#' T <- 10
+#' K <- length(means)
+#' Bandit <- ExploreFristNBandit$new(K, T, N)
+#' for (t in seq(T)) {
+#'     # get which arm to choose
+#'     k <- Bandit$value$which
+#'     # update the streamer with a value drawn from the kth arm
+#'     Bandit$update(rnorm(1, mean = means[k]), k)
+#' }
+#'
+#' @export
+#' @format An \code{\link{R6Class}} generator object
+ExploreFristNBandit <- R6::R6Class(
+    "ExploreFirstNBandit",
+    inherit = MultiArmedBandit,
+    public = list(
+        #' @description Creates a new \code{ExploreFirstNBandit} streamer
+        #' object.
+        #'
+        #' @param K the number of arms
+        #' @param T number of rounds / draws
+        #' @param N number of draws in the exploration phase, if not specified
+        #' the optimal value for N = (T/K) ^ (2/3) * log(T) ^ (1/3) is chosen.
+        #'
+        #' @examples
+        #' Bandit <- ExploreFristNBandit$new(3, 20, 2)
+        #'
+        #' @return The new \code{ExploreFristNBandit} (invisibly)
+        initialize = function(K, T, N = NULL) {
+            super$initialize(K, T)
+            private$state <- "exploration"
+            if (is.null(N)) {
+                N <- floor((T / K) ^ (2 / 3) * log(T) ^ (1 / 3))
+            }
+            if (T < N * K) {
+                stop(sprintf(
+                    "More draws are required for the values of
```

Comment:
I think this error could be clearer. Something like "More exploration draws are specified than the total number of rounds"